demos/mantle_convection/adaptive_base_case/adaptive_base_case.py

```diff
@@ -0,0 +1,362 @@
+# as explained
+# [here](https://github.com/mesh-adaptation/docs/wiki/Installation-Instructions).
+# Additionally, for the assembly of the metric field, which allows us to
+# specify exactly where mesh resolution is needed, we use the
+# [animate](https://mesh-adaptation.github.io/docs/animate/index.html) python
+# package that can be pip installed from
+# https://github.com/mesh-adaptation/animate
+
+from gadopt import *
+from animate import RiemannianMetric, adapt
+
+# Set up the initial mesh and timestepping options
+
+nx, ny = 10, 10
+mesh = UnitSquareMesh(nx, ny, quadrilateral=False)  # Square mesh generated via firedrake
+
+time = 0.0  # Initial time
+timestep = 0  # Initial timestep
+timesteps_per_adapt = 10
+delta_t = Constant(1e-6)  # Initial time-step
+
+# Create pvd-file to output solutions fields at the specified
+# output frequency, which we can visualise using paraview
+# or pyvista. We need the additional `adaptive=True` as
+# the mesh will not be the same during the entire simulation.
+# We choose the same output frequency as the number of timesteps
+# between mesh adapts, so that we get one output on each different
+# mesh
+# Additionally, we open a log file to output diagnostic values.
+
+# +
+output_file = VTKFile("output.pvd", adaptive=True)
+output_frequency = timesteps_per_adapt  # Output every adapt
+
+plog = ParameterLog('params.log', mesh)
+plog.log_str("timestep time dt maxchange u_rms u_rms_surf ux_max nu_top nu_base energy avg_t")
+# -
+
+# Initial conditions for the model, these will be interpolated onto
+# the initial mesh later on.
+
+X = SpatialCoordinate(mesh)
+T_init = 1.0 - X[1] + 0.05 * cos(pi * X[0]) * sin(pi * X[1])
+u_init = as_vector((0., 0.))
+p_init = 0.
+
+# In all other demos, we would now continue defining some function spaces,
+# create some functions, and then the solvers using those. As we will be
+# adapting the mesh these steps will need to be repeated after each adapt.
+# Therefore we will wrap these steps in a python function. The following python
+# function in fact wraps the entire base case: it takes in a mesh builds up the
+# function spaces, functions and solvers and then runs the model for a
+# specified number of timesteps. We will be using this function in an outside
+# loop, where we adapt the mesh and then call the function to run the model on
+# the adapted mesh for a number of timesteps. The temperature, velocity
+# and pressure solution at the end of those timesteps, which we have solved
+# on the current adapted mesh, will need to be interpolated after the next mesh
+# adaptivity stage onto the new mesh and are therefore provided as `T_init`,
+# `u_init`, and `p_init` arguments. In the first call to this subroutine, the mesh
+# is still the initial mesh, and we simply use the initial condition expressions
+# we have just defined.
+
+# +
+
+
+def run_interval(mesh, time, timestep, Nt, T_init, u_init, p_init):
```

Comment:
Perhaps bring this whole function and the commented block above to the top of the demo. This way, you can preserve the continuity of the rest of the code blocks.